src/engine.ml

```diff
@@ -584,51 +594,58 @@ let incoming state now ts buf =
                set_ch state ch, out, act
            end
          | (`Ack _ | `Control _) as d ->
-           check_control_integrity bad_mac key p state.session.their_hmac >>= fun () ->
-           (* _first_ update state with most recent received packet_id *)
-           expected_packet state.session ch.transport d >>= fun (session, transport) ->
-           let state' = { state with session }
-           and ch' = { ch with transport }
-           in
-           match d with
-           | `Ack _ ->
-             (* TODO nothing to do for an ack, should bump ack number *)
-             Logs.info (fun m -> m "ignoring acknowledgement");
-             Ok (set_ch state' ch', out, act)
-           | `Control (typ, (_, _, data)) ->
-             incoming_control state.config state.rng state'.session ch' now typ data
-             >>| fun (est, config', ch'', outs) ->
-             Logs.debug (fun m -> m "out channel %a, pkts %d"
-                            pp_channel ch'' (List.length outs));
-             (* each control needs to be acked! *)
-             let outs = match outs with [] -> [ Cstruct.empty ] | xs -> xs in
-             (* now prepare outgoing packets *)
-             let session, transport, encs =
-               wrap_hmac_control now state'.session key ch''.transport outs
+           match
+             check_control_integrity bad_mac key p state.session.their_hmac >>= fun () ->
+             (* _first_ update state with most recent received packet_id *)
+             expected_packet state.session ch.transport d
+           with
+           | Error e ->
+             (* only in udp mode? *)
```

Comment:
If we do this we need constant-time comparison in `check_control_integrity` to prevent timing-assisted brute force attacks from bypassing the MAC check. Not sure if we care here?